composer/datasets/in_context_learning_evaluation.py

```diff
@@ -103,117 +108,171 @@ def _get_fewshot_sample_idxs(dataset_size: int, num_fewshot: int, sample_idx: in
+                'prompt_string': prompt_string,
+                'fewshot_rng': fewshot_rng,
+            },
+        )
 
-    def __init__(
+    def __getitem__(self, index: int):
+        return self.encoded_dataset[index]
+
+    def __len__(self):
+        return len(self.encoded_dataset)
+
+    def get_num_samples_in_batch(self, batch: dict) -> int:
+        return batch['input_ids'].shape[0]
+
+    def _read_dataset(
         self,
         dataset_uri: str,
-        tokenizer: Union[transformers.PreTrainedTokenizer, transformers.PreTrainedTokenizerFast],
-        max_seq_len: int,
-        pad_tok_id: int,
-        num_fewshot: int,
-        prompt_string: str,
-        example_delimiter: str,
-        continuation_delimiter: str,
         destination_path: str,
-        question_prelimiter: str,
-        fewshot_random_seed: int,
-        cot_delimiter: str = '',
+        hf_loading_vars: dict = None,
+        hf_parsing_vars: dict = None,
     ):
         try:
             from datasets import load_dataset  # pyright: ignore [reportGeneralTypeIssues]
         except ImportError as e:
-            raise MissingConditionalImportError(extra_deps_group='nlp',
-                                                conda_package='datasets',
-                                                conda_channel='conda-forge') from e
-        with dist.local_rank_zero_download_and_wait(destination_path):
-            if dist.get_local_rank() == 0:
-                get_file(dataset_uri, destination_path, overwrite=True)
-        dataset = load_dataset('json', data_files=destination_path, split='train', streaming=False)
-        self.samples = self._read_dataset(dataset)
-        self.samples = strip_data(self.samples)
-        self.tokenizer = tokenizer
-        self.max_seq_len = max_seq_len
-        self.pad_tok_id = pad_tok_id
-        self.padding_side = 'left'
-        self.max_answer_length = 0
-        fewshot_rng = random.Random(fewshot_random_seed)
-        self.encoded_dataset = self._prep_examples(num_fewshot, prompt_string, example_delimiter,
-                                                   continuation_delimiter, question_prelimiter, fewshot_rng,
-                                                   cot_delimiter)
-
-    def _format_prompt_and_fewshot(self, num_fewshot: int, prompt_string: str, example_delimiter: str,
-                                   continuation_delimiter: str, question_prelimiter: str, cot_delimiter: str,
-                                   fewshot_rng: random.Random, sample_idx: int) -> str:
+            raise MissingConditionalImportError(
+                extra_deps_group='nlp',
+                conda_package='datasets',
+                conda_channel='conda-forge',
+            ) from e
+        if 'hf://' in dataset_uri:
+            dataset_uri = dataset_uri.replace('hf://', '')
+            dataset = load_dataset(dataset_uri, **hf_loading_vars)
+            batched = hf_parsing_vars.pop('batched', False)
```

Comment:
why are the parsing vars popped but the loading vars **? **is better cause otherwise you can have typos in the parsing vars.

Also, should we support batched parsing? seems like maybe fine/simpler not to?

Reply:
MUCH simpler not to. I'd love to not support batching, but in the case where one row of the HF dataset should be parsed out into many rows in the icl dataset,  [you need batched=True](https://huggingface.co/docs/datasets/v2.14.7/en/package_reference/main_classes#datasets.Dataset.map:~:text=the%20following%20signatures%3A-,function(example%3A%20Dict%5Bstr%2C%20Any%5D)%20%2D%3E%20Dict%5Bstr%2C%20Any%5D%20if%20batched,True%20and/or%20with_rank%3DTrue%20(one%20extra%20arg%20for%20each),-For%20advanced%20usage) (for example in leval, they have a single row containing one col of 15 questions, 15 answers, and one document, such that all the q/a pairs are about that document. You'd want to parse that into 15 total rows). 

I popped it out of `hf_parsing_vars` bcus it's a variable for parsing the hf dataset and thus belonged there in the yaml, even if it's not passed into the defined function directly.

Reply:
this is no longer relevant as we will not support batching